gateway/ln-gateway/src/state_machine/pay.rs

```diff
@@ -242,21 +264,75 @@ impl GatewayPayInvoice {
         }
     }
 
+    async fn buy_preimage_via_direct_swap(
+        client: Client,
+        invoice: Bolt11Invoice,
+        contract: OutgoingContractAccount,
+    ) -> Result<(FederationId, OperationId), OutgoingPaymentError> {
+        let swap_params = invoice
+            .try_into()
+            .map_err(|e| OutgoingPaymentError::SwapFailed {
+                contract: contract.clone(),
+                swap_error: format!("Failed to parse invoice into swap parameters: {}", e),
+            })?;
+
+        let operation_id = client
+            .gateway_handle_direct_swap(swap_params)
+            .await
+            .map_err(|e| OutgoingPaymentError::SwapFailed {
+                contract: contract.clone(),
+                swap_error: format!("Failed to initiate direct swap: {}", e),
+            })?;
+
+        Ok((client.federation_id(), operation_id))
+    }
+
     async fn transition_buy_preimage(
         context: GatewayClientContext,
         result: Result<(OutgoingContractAccount, PaymentParameters), OutgoingPaymentError>,
         common: GatewayPayCommon,
     ) -> GatewayPayStateMachine {
         match result {
             Ok((contract, payment_parameters)) => {
-                let preimage_result = Self::buy_preimage_over_lightning(
+                if let Some(client) = Self::check_swap_to_federation(
+                    context.clone(),
+                    payment_parameters.invoice.clone(),
+                )
+                .await
+                {
+                    match Self::buy_preimage_via_direct_swap(
+                        client,
+                        payment_parameters.invoice.clone(),
+                        contract.clone(),
+                    )
+                    .await
+                    {
+                        Ok((federation_id, operation_id)) => {
+                            return GatewayPayStateMachine {
+                                common,
+                                state: GatewayPayStates::WaitForSwapPreimage(Box::new(
+                                    GatewayPayWaitForSwapPreimage {
+                                        contract,
+                                        federation_id,
+                                        operation_id,
+                                    },
+                                )),
+                            };
+                        }
+                        Err(e) => {
+                            tracing::debug!("Failed to buy preimage via direct swap: {}", e)
+                            // Fall back to lightning payments
```

Comment:
Hmm do we want to fall back here? If the node in the route hint matches the gateway node, that means that the lightning node is always going to try and do a self payment, which we know won't work. Maybe we should just fail?

Reply:
I was undecided on the desired behavior. I agree failing here makes more sense